runtime/expr/function/grok.go

```diff
@@ -0,0 +1,110 @@
+package function
+
+import (
+	"errors"
+	"regexp"
+	"strings"
+
+	"github.com/brimdata/zed"
+	"github.com/brimdata/zed/pkg/grok"
+	"github.com/brimdata/zed/zcode"
+	"golang.org/x/exp/slices"
+)
+
+type Grok struct {
+	zctx    *zed.Context
+	builder zcode.Builder
+	hosts   map[string]*host
+}
+
+func newGrok(zctx *zed.Context) *Grok {
+	return &Grok{
+		zctx:  zctx,
+		hosts: make(map[string]*host),
+	}
+}
+
+func (g *Grok) Call(ectx zed.Allocator, vals []zed.Value) *zed.Value {
+	v, err := g.call(ectx, vals)
+	if err != nil {
+		return ectx.CopyValue(*g.zctx.NewErrorf("grok(): %s", err))
+	}
+	return v
+}
+
+func (g *Grok) call(ectx zed.Allocator, vals []zed.Value) (*zed.Value, error) {
+	extraArg, patternArg, inputArg := zed.NullString, vals[0], vals[1]
+	if len(vals) == 3 {
+		extraArg, patternArg, inputArg = &vals[0], vals[1], vals[2]
+	}
+	switch {
+	case zed.TypeUnder(extraArg.Type) != zed.TypeString:
+		return nil, errors.New("external pattern argument must be a string")
+	case zed.TypeUnder(patternArg.Type) != zed.TypeString:
+		return nil, errors.New("pattern argument be a string")
+	case zed.TypeUnder(inputArg.Type) != zed.TypeString:
+		return nil, errors.New("input argument be a string")
+	}
+	h, err := g.getHost(extraArg.AsString())
+	if err != nil {
+		return nil, err
+	}
+	p, err := h.getPattern(g.zctx, patternArg.AsString())
+	if err != nil {
+		return nil, err
+	}
+	ss := p.ParseValues(inputArg.AsString())
+	if ss == nil {
+		return nil, errors.New("no match")
```

Comment:
Instead of just `error("grok(): no match")`, might we consider wrapping the original input string in the `error` value? This way if a user has set up a `zq` Grok pipeline output to `zed load` where they're expecting all the input data to be Grok-able, and some un-parseable data shows up in the input stream, they'll still have the option to find the original data inside the error and parse/load it after they've updated their Grok pattern to catch it going forward. This feels like it's in line with some of our "why Zed errors are great" stuff that's covered in places like [this blog post](https://www.brimdata.io/blog/debugging-data-first-class-errors/). It's also kind of similar to Logstash's [`_grokparsefailure` tag](https://www.elastic.co/guide/en/logstash/8.10/plugins-filters-grok.html#plugins-filters-grok-tag_on_failure) that they add on parse failures. Their overall approach is a little different than ours since by default they always include the original input string in the output as a field called `message` (see the many Logstash outputs in my recent [big comment](https://github.com/brimdata/zed/pull/4827#issuecomment-1788123901)) but it would seem to enable a similar "repair after the fact" workflow.